### ISSUE TITLE
Non-existent POSIX_MADV_FREE macro and wrong usage of POSIX_MADV_DONTNEED macro

### ISSUE THREAD

**Issue by praetoriandroid** [mjansson/rpmalloc#210]
Hello!

With the commit 09f4ca1bc515fe1862a1426970177c9d0b340829 (Port to Haiku (#76)) ```madvise()``` calls in the rpmalloc.c were replaced with ```posix_madvise()``` and ```MADV_*``` constants were replaced with ```POSIX_MADV_*``` ones. But there is no POSIX_MADV_FREE constant in Haiku or in any other OS, so the code
```C
#if defined(POSIX_MADV_FREE)
		if (posix_madvise(address, size, POSIX_MADV_FREE))
#endif
```
just does nothing on any platform.

Also, the ```POSIX_MADV_DONTNEED``` is not the same as ```MADV_DONTNEED```. At least on GNU/Linux and on Android.

From ```man posix_madvise```:
> In glibc, this function is implemented using madvise(2).  However, since  glibc  2.6,  POSIX_MADV_DONTNEED  is
> treated as a no-op, because the corresponding madvise(2) value, MADV_DONTNEED, has destructive semantics.

From the Bionic libc source code (Android):
```C
int posix_madvise(void* addr, size_t len, int advice) {
  ErrnoRestorer errno_restorer;

  // Don't call madvise() on POSIX_MADV_DONTNEED, it will make the space not available.
  if (advice == POSIX_MADV_DONTNEED) {
    return 0;
  }
  return (madvise(addr, len, advice) == 0 ? 0 : errno);
}
```
So on these platforms ```posix_madvise(address, size, POSIX_MADV_DONTNEED)``` does nothing either.

Furthermore, in commit c6680f548cb5762b2c9a89f2427cf38b864a1f13 (On Darwin using MADV_FREE_REUSABLE to notify the system (#208)) it was added one more call to ```madvise()``` before existing two. If I understood correctly, ```MADV_FREE_REUSABLE``` is something like a better version of ```MADV_FREE``` on macOS and iOS.

Also, if I understood right, you want to call ```posix_madvise(address, size, POSIX_MADV_DONTNEED)``` only if ```posix_madvise(address, size, POSIX_MADV_FREE)``` was failed.

Probably, if ```posix_madvise(..., POSIX_MADV_FREE)``` will be reverted to ```madvise(..., MADV_FREE)```, there will be no need to use both ```MADV_FREE``` and ```MADV_FREE_REUSABLE``` simultaneously. And depending on existence and meaning of ```MADV_DONTNEED``` and ```POSIX_MADV_DONTNEED``` on various platforms, something like the next patch may be applied:
```diff
diff --git a/rpmalloc/rpmalloc.c b/rpmalloc/rpmalloc.c
index da3f75e..cd5f7d9 100644
--- a/rpmalloc/rpmalloc.c
+++ b/rpmalloc/rpmalloc.c
@@ -868,11 +868,14 @@ _rpmalloc_unmap_os(void* address, size_t size, size_t offset, size_t release) {
 	} else {
 #if defined(MADV_FREE_REUSABLE)
 		if (madvise(address, size, MADV_FREE_REUSABLE))
+#elif defined(MADV_FREE)
+		if (madvise(address, size, MADV_FREE))
 #endif
-#if defined(POSIX_MADV_FREE)
-		if (posix_madvise(address, size, POSIX_MADV_FREE))
-#endif
+#if defined(MADV_DONTNEED)
+		if (madvise(address, size, MADV_DONTNEED)) {
+#else
 		if (posix_madvise(address, size, POSIX_MADV_DONTNEED)) {
+#endif
 			assert("Failed to madvise virtual memory block as free" == 0);
 		}
 	}
```
So that the resulting code will be:
```C
#if defined(MADV_FREE_REUSABLE)
		if (madvise(address, size, MADV_FREE_REUSABLE))
#elif defined(MADV_FREE)
		if (madvise(address, size, MADV_FREE))
#endif
#if defined(MADV_DONTNEED)
		if (madvise(address, size, MADV_DONTNEED)) {
#else
		if (posix_madvise(address, size, POSIX_MADV_DONTNEED)) {
#endif
			assert("Failed to madvise virtual memory block as free" == 0);
		}
```

P.S. By the way, the support for ```madvise()``` and even ```MADV_FREE``` extension [has been added to Haiku](https://github.com/haiku/haiku/commit/b3bd66961a8cbdf76428bf78f0e2df783219dd79).

**Comment by mjansson** [comment]
Thank you for the detailed info, will look into this asap

**Comment by mjansson** [comment]
This was indeed a mistake and your suggested changes looks good. Fixed in e92611d
